src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs

```diff
@@ -65,50 +64,57 @@ internal abstract partial class AbstractAsynchronousTaggerProvider<TTag> : Foreg
         protected virtual IEnumerable<Option2<bool>> Options => SpecializedCollections.EmptyEnumerable<Option2<bool>>();
         protected virtual IEnumerable<PerLanguageOption2<bool>> PerLanguageOptions => SpecializedCollections.EmptyEnumerable<PerLanguageOption2<bool>>();
 
+        protected virtual bool ComputeInitialTagsSynchronously(ITextBuffer subjectBuffer) => false;
+
+        /// <summary>
+        /// How long the tagger should wait after hearing about an event before recomputing tags.
+        /// </summary>
+        protected abstract TaggerDelay EventChangeDelay { get; }
+
         /// <summary>
         /// This controls what delay tagger will use to let editor know about newly inserted tags
         /// </summary>
         protected virtual TaggerDelay AddedTagNotificationDelay => TaggerDelay.NearImmediate;
 
-        protected virtual bool ComputeInitialTagsSynchronously(ITextBuffer subjectBuffer) => false;
-
 #if DEBUG
         public readonly string StackTrace;
 #endif
 
-        protected AbstractAsynchronousTaggerProvider(
-            IThreadingContext threadingContext,
-            IAsynchronousOperationListener asyncListener,
-            IForegroundNotificationService notificationService)
+        protected AbstractAsynchronousTaggerProvider(IThreadingContext threadingContext, IAsynchronousOperationListener asyncListener)
             : base(threadingContext)
         {
             AsyncListener = asyncListener;
-            _notificationService = notificationService;
 
 #if DEBUG
             StackTrace = new StackTrace().ToString();
 #endif
         }
 
-        /// <summary>
-        /// How long the tagger should wait after hearing about an event before recomputing tags.
-        /// </summary>
-        protected abstract TaggerDelay EventChangeDelay { get; }
-
-        internal ITagger<T>? CreateTaggerWorker<T>(ITextView textViewOpt, ITextBuffer subjectBuffer) where T : ITag
+        protected ITagger<T>? CreateTaggerWorker<T>(ITextView textViewOpt, ITextBuffer subjectBuffer) where T : ITag
         {
             if (!subjectBuffer.GetFeatureOnOffOption(EditorComponentOnOffOptions.Tagger))
                 return null;
 
             var tagSource = GetOrCreateTagSource(textViewOpt, subjectBuffer);
-            return new Tagger(ThreadingContext, AsyncListener, _notificationService, tagSource, subjectBuffer) as ITagger<T>;
+            var tagger = new Tagger(tagSource);
+
+            // If we're not able to convert the tagger we instantiated to the type the caller wants, then make sure we
+            // dispose of it now.  The tagger will have added a ref to the underlying tagsource, and we have to make
+            // sure we return that to the property starting value.
+            if (tagger is not ITagger<T> result)
```

Comment:
How is this possible? Do generic type parameters have weird behaviour over nested classes or something?

Reply:
So it's very possible in that this method is a generic method in the non generic ITaggerProvider interface. So a caller may mef import all the taggers, and then try to create a tagger with a type parameter we are not providing. 

As a simple example, imagine they get our outlining tagger provider, and they ask it to create an ITagger<ClassificationTag>. This won't work at all. 

This is caused by a bug in the caller. When they mef imported us, that should have seen what type of tags we say we supply in our exports, and they should only try to instantiate to that. But that's a contract the type system can't enforce, so we need this code to prevent a leak if the caller does this screwy thing.

Reply:
Thanks, completely missed that this was a generic method, I read the check as `ITagger<TTag>`